stacktrace.go

```diff
@@ -7,7 +7,20 @@ import (
 	"github.com/palantir/witchcraft-go-error/internal/errors"
 )
 
-func callers() *stack {
+var _ StackTrace = (*stack)(nil)
+
+// StackTrace provides formatting for an underlying stack trace.
+type StackTrace interface {
+	fmt.Formatter
+}
+
+// StackTracer provides the behavior necessary to retrieve a StackTrace formatter.
+type StackTracer interface {
+	StackTrace() StackTrace
+}
+
+// NewStackTrace creates a new StackTrace, constructed by collecting program counters from runtime callers.
+func NewStackTrace() StackTrace {
```

Comment:
now that this is exported, it might be called from utilities that add an arbitrary number of extra stack frames. Maybe we want a `skip int` argument and below do `4+skip` or whatever the math ends up being. That way the wrapping functions don't show up in the stacktraces

Reply:
I prefer the cleaner function for most users, so I created a second helper `NewStackTraceWithSkip` that takes `skip int` as an argument. Unfortunately, this requires duplicating the logic because having `NewStackTrace` call into `NewStackTraceWithSkip` adds an additional frame. 

I'm open to going with your suggestion of having a single function that takes a skip value if you feel strongly. Otherwise, I could also just leave this out until someone has a use case for it.

Reply:
Now that it's exposed, this one can become

```go
func NewStackTrace() StackTrace {
    return NewStackTraceWithSkip(1)
}
```
```

Reply:
lol, yes, correct